libraries/chain/block_state.cpp

```diff
@@ -88,7 +88,9 @@ std::pair<vote_status, std::optional<uint32_t>> block_state::aggregate_vote(cons
                                  vote.finalizer_key,
                                  vote.sig,
                                  finalizers[index].weight);
-      return {status, strong ? core.final_on_strong_qc_block_num : std::optional<uint32_t>{}};
+      return {status,
+              (status == vote_status::success && strong) ?
+                 core.final_on_strong_qc_block_num : std::optional<uint32_t>{}};
```

Comment:
When it returns a non-null-optional, the caller will go ahead to call `set_if_irreversible_block_num` which is a waste. I also think it is cleaner: if a vote is bad, both return values (`status` and `strong`) should be set properly so that they cannot be used. @heifner ?

Reply:
In my branch I have this as:
```cpp
      std::optional<uint32_t> new_lib{};
      if (status == vote_status::success && state == pending_quorum_certificate::state_t::strong)
         new_lib = core.final_on_strong_qc_block_num;
      return {status, state, new_lib};

```

Reply:
I'm fine with Kevin's version.

Reply:
Honestly, I liked it better before any change. The meaning was clear. We're asking `block_state` to aggregate a vote into its `pending_qc`, and it returns a pair where:

`pair.first` is the status of applying the vote (so we know whether to propagate the vote)
`pair.second` is the new `lib` that `pending_qc` certifies (if it achieved strong status). The caller decides what to do with this info.

I'm not understanding why it helps to make it more complicated than that.

Reply:
This is changing in my PR. We have to let fork_db tell us new lib.

Reply:
Isn't still fine to let `aggregate_vote` tell you which block its `pending_qc` is making final, even if that's not the lib that the controller will end up storing?

Reply:
Potentially, but as is it is not telling you the correct answer because it is not correctly taking previous votes into account. It is only looking at `core` not `updated_core`.

Reply:
I'm not understanding what `updated_core` is about? I thought you may need to update the core of other block state (the descendants), but why the one we are voting on?